### ISSUE TITLE
Consuming event bus events

### ISSUE THREAD

**Issue by robrap** [openedx/openedx-events#59]
In the ADR (https://github.com/openedx/openedx-events/blob/main/docs/decisions/0004-external-event-bus-and-django-signal-events.rst), we discussed firing the same event signal after consuming an event bus event.

An advantage of this proposal is it mimics what is done for in-process events.

However, there is a major difference between in-process events and events in the consuming application. For in-process events, we must use a technology like celery to get asynchronous processing of the events. However, our event bus consuming applications are already asynchronous in relation to the main user/api-facing service.

For these applications, we have an opportunity to simply do the event processing in-process from the point of view of the consuming application, and to horizontally scale the consumer containers based on processing requirements. We could do this without spinning up separate celery tasks that will then compete with other tasks in the same celery queue.

* Should each consumer application only handle one type of processing per event within itself?
* Are signals still a good pattern for making the processing pluggable, or does it lead one to assume that multiple receivers may be configured? Are we ok with multiple signal receivers?
* If multiple signal receivers are used, would we have recommendations on whether or not some or all of them are allowed to processing within the consumer, vs spinning up a separate task in celery?
* If we want only one processor per consumer, should we make a different pluggable interface that is more like a singleton receiver?

Note, I personally would like to take advantage of processing in the consumers and avoiding celery to start. We don't need to make any final decisions here, but it would be nice to have a nice initial recommendation.

Thank you.

**Comment by feanil** [comment]
@robrap I think at the very least, with the way openedx-events is setup right now we'd get multiple consumers of events anyway right? Because there is a chance that various in-process plugins want to consume the event?  I agree that adding celery to the mix seems overkill if all we're doing is serializing and pushing it to a datastore.  We do this all the time in-process for other data stores, I think it's fine to do so again for event-bus purposes since those backend stores are still pretty fast for writes most of the time(tm).

**Comment by robrap** [comment]
Thanks for the response @feanil. Let me clarify some things:
* We will definitely have a different consumer per topic (related event types) per service (e.g. lms, credentials, discovery).
* However, a single service (e.g. lms) may want to do multiple things as a result of an event bus event (just like it sometimes wants to do multiple things based on an in-process signal based event). For example, let's say the LMS has two plugins that each want to do something in the LMS service based on the same event bus event. Each plugin could be required to have its own cluster of consumer containers, so it has no competition when processing the events. Or, each plugin might register signal receivers in the same way it does for in-process events, but their they are asked to spin off celery tasks to actually do the work. Here, would they each just do the work in the same consumer? If one plugin had a very small amount of work, and the other a large amount of work, that wouldn't be fair.
* Same issue for any core (non-plugin) functionality for processing an event bus event, and it potentially competing with other receivers.

Writing this answer, and agreeing we don't want to have to use celery, I hope we can simply limit one receiver/event processor per consumer and make it extensible, so a plugin or core code could configure its unique consumer group and k8s cluster (in our deployment).

Lastly, although handling an event may be as simple as pushing an update to a datastore, it might also kick-off a whole new process or report execution, or anything, so I don't want to assume it is light-weight.

A simple alternative is that any plugin to the same service would process under the same consumer group (topic+service), and unlike events in the home service, would simply execute whatever it needs to do, and we punt on this issue until when and if we have competing processing with competing durations and priorities, which may never be the case. Am I thinking too far ahead?

**Comment by ormsbee** [comment]
I'm inclined to punt the problem to consumers for now, with the assumption that the default approach will be for consumers to process a message synchronously, but have the option to spin up a celery worker for something particularly long running. It requires the least development effort, and is very close to what's already being done on the web side.

I like the idea of removing celery from the equation in principle, and I think we'll get there eventually. But I don't think we know enough about the problems we're trying to solve right now, and I'm worried that we'll over-engineer a solution. Other concerns for the short term:

* **Monitoring/error reporting gap**
We know how to do this for celery processes, but would need to build a new thing here.
* **Idle processes**
A lot of our events are low throughput, and plugins might only operate on a very small subset of events, or do only some trivially simple operation on them. In that event, we'd have operational complexity without as much benefit.
* **Configuration Complexity**
I just have this morbid vision of multiple RCAs that start with, "Why didn't thing Foo receive event Bar, and why didn't we notice it wasn't connected?"

So I think we should opt to have larger consumers that:

1. Connect to the event bus and receive the event.
2. Use openedx-events to turn that event into the proper signal.
3. Send that signal within the consumer process.
4. Let the consumer decide whether it can do things in-process or it needs to spin up a celery task for something long-running.

**Comment by robrap** [comment]
@ormsbee: Thanks for the detailed response. Sorry for the delay in getting back to you.

> 4. Let the consumer decide whether it can do things in-process or it needs to spin up a celery task for something long-running.

Makes sense and agreed. It is basically in-line with what I was thinking. I was less certain about long-running processes, because people had expressed some ideal of moving away from celery, but I am now convinced we need something other than the consumer for that, and at this point, that is celery.

That said, I'm a little confused about the context for the bullets under:

> Other concerns for the short term: ...

I guess I see your point around "Idle processes", but it feels like monitoring, and avoiding "Why didn't thing Foo receive event Bar, and why didn't we notice it wasn't connected?", both come with the territory of the event bus no matter what direction we go. I am probably missing something, or maybe you thought these would just get even harder for some reason if we did something more complex? Anyway, feel free to explain if you think I need to better understand.

**Comment by ormsbee** [comment]
> I guess I see your point around "Idle processes", but it feels like monitoring, and avoiding "Why didn't thing Foo receive event Bar, and why didn't we notice it wasn't connected?", both come with the territory of the event bus no matter what direction we go. I am probably missing something, or maybe you thought these would just get even harder for some reason if we did something more complex? Anyway, feel free to explain if you think I need to better understand.

My concern was mostly just around adding points of failure where things could go wrong, as we're going to be dealing with a lot of new moving parts. I figured that if we did the more ambitious version of this where each plugin can have its own set of processes, then we'd need to somehow configure scaling rules for that, and make sure all the plugins were being monitored (vs. the one process), and that events were mapped to individual plugins properly. In terms of monitoring, we already have stuff around celery (like flower), and New Relic tracking of exceptions and workers, etc.

It might be the case that there are simple equivalents to this, and I'm just not seeing how it would fit together. I honestly haven't thought about it for long enough to have a strong opinion on it–just long enough to say, "yeah, that sounds harder than it's worth for the moment". 😛
